extensions/zacoons/webgpu.js

```diff
@@ -0,0 +1,638 @@
+      return this.bindGroups.length - 1;
+    }
+
+    // compute pipelines
+    createComputePipeline({ LAYOUT, MODULE_NAME }) {
+      this.pipelines.push(
+        this.dev.createComputePipeline({
+          layout: this.pipelineLayouts[LAYOUT],
+          compute: {
+            module: this.modules[MODULE_NAME],
+            entryPoint: "main",
+          },
+        })
+      );
+      return this.pipelines.length - 1;
+    }
+
+    // command buffers
+    newCommandBuffer() {
+      this.tmp = {};
+      this.tmp.commandEncoder = this.dev.createCommandEncoder();
+    }
+    beginComputePass({ PIPELINE }) {
+      this.tmp.computePass = this.tmp.commandEncoder.beginComputePass();
+      this.tmp.pipeline = this.pipelines[PIPELINE];
+      this.tmp.computePass.setPipeline(this.tmp.pipeline);
+    }
+    setComputePassBindGroup({ BIND_GROUP_IDX, BIND_GROUP }) {
+      this.tmp.computePass.setBindGroup(
+        BIND_GROUP_IDX,
+        this.bindGroups[BIND_GROUP]
+      );
+    }
+    dispatchComputePass({ X, Y, Z }) {
+      this.tmp.computePass.dispatchWorkgroups(X, Y, Z);
+      this.tmp.computePass.end();
+    }
+    copyBufferToBuffer({ SRC, DST }) {
+      const src = this.buffers[SRC];
+      this.tmp.commandEncoder.copyBufferToBuffer(
+        src,
+        0,
+        this.buffers[DST],
+        0,
+        src.size
+      );
+    }
+    createCommandBuffer() {
+      this.commandBuffers.push(this.tmp.commandEncoder.finish());
+      return this.commandBuffers.length - 1;
+    }
+
+    // queues
+    newQueue() {
+      this.tmp = [];
+    }
+    writeListToBuffer({ LIST_ID, TYPE, BUFFER }) {
+      this.dev.queue.writeBuffer(
+        this.buffers[BUFFER],
+        0,
+        this.types[TYPE].from(this._getList(LIST_ID))
+      );
+    }
+    addCommandBufferToQueue({ COMMAND_BUFFER }) {
+      this.tmp.push(this.commandBuffers[COMMAND_BUFFER]);
+    }
+    submitQueue() {
+      this.dev.queue.submit(this.tmp);
+    }
+
+    // helpers
+    _resetAttrs() {
+      this.dev?.destroy();
+      this.dev = undefined;
+      for (const b of this.buffers || []) b.destroy();
+
+      this.modules = {};
+      this.bindGroupLayouts = [];
+      this.pipelineLayouts = [];
+      this.buffers = [];
+      this.bindGroups = [];
+      this.pipelines = [];
+      this.commandBuffers = [];
+
+      /** @type {any} */
+      this.tmp = undefined;
+    }
+    _getModuleNames() {
+      const names = [];
+      for (const commentName in Scratch.vm.editingTarget.comments) {
+        const c = Scratch.vm.editingTarget.comments[commentName];
+        if (c.text.startsWith("name:")) {
+          const firstLineEndIdx = c.text.indexOf("\n");
+          names.push(c.text.slice(5, firstLineEndIdx).trim());
+        }
+      }
+      return names.length ? names : [""];
+    }
+    _getLists() {
+      const names = [];
+
+      const globalVars = Object.entries(
+        Scratch.vm.runtime.targets[0].variables

```

Comment:
It's better to use vm.runtime.getTargetForStage() which finds stage regardless of where it is in targets array.